### PR TITLE
Create CF template subnets from multiple AZs

### PR DESCRIPTION
**Pull request by tuommaki**
Each availability zone requires separate public/private subnets. These
are created here.

This change is not working alone, but requires also similar changes for
routing tables, nat gateways and peering routes.

### CODE REVIEW

File: service/controller/v19/adapter/guest_subnets.go [hunk 1]
```diff
@@ -1,63 +1,44 @@
...
 func (s *GuestSubnetsAdapter) Adapt(cfg Config) error {
-	publicSubnet, privateSubnet, err := allocatePublicAndPrivateSubnets(cfg)
-	if err != nil {
-		return microerror.Mask(err)
+	for i, az := range key.AvailabilityZones(cfg.CustomObject) {
+		snet := Subnet{
+			Index:               i,
+			AvailabilityZone:    az.Name,
+			CIDR:                az.Subnet.Public.CIDR,
+			Name:                fmt.Sprintf("PublicSubnet%d", i),
```
Comment (kopiczko) [inline service/controller/v19/adapter/guest_subnets.go]:
```suggestion
			Name:                fmt.Sprintf("PublicSubnet%02d", i),
```

I think it's a bit cooler when they will have the same length.

File: service/controller/v19/adapter/guest_subnets.go [hunk 2]
```diff
@@ -1,63 +1,44 @@
...
+			MapPublicIPOnLaunch: false,
+		}
+		s.PublicSubnets = append(s.PublicSubnets, snet)
+
+		snet = Subnet{
+			Index:               i,
+			AvailabilityZone:    az.Name,
+			CIDR:                az.Subnet.Private.CIDR,
+			Name:                fmt.Sprintf("PrivateSubnet%d", i),
```
Comment (kopiczko) [inline service/controller/v19/adapter/guest_subnets.go]:
```suggestion
			Name:                fmt.Sprintf("PrivateSubnet%02d", i),
```

File: service/controller/v19/adapter/guest_subnets.go [hunk 3]
```diff
@@ -1,63 +1,44 @@
...
-	PublicSubnetMapPublicIPOnLaunch  bool
-	PrivateSubnetAZ                  string
-	PrivateSubnetCIDR                string
-	PrivateSubnetName                string
-	PrivateSubnetMapPublicIPOnLaunch bool
+	PublicSubnets  []Subnet
+	PrivateSubnets []Subnet
 }
 
 func (s *GuestSubnetsAdapter) Adapt(cfg Config) error {
```
Comment (kopiczko) [inline service/controller/v19/adapter/guest_subnets.go]: I'd sort the AZs here instead in `key.AvailabilityZones`. It's kind of indirect to rely on that.

File: service/controller/v19/key/key.go [hunk 4]
```diff
@@ -111,8 +112,12 @@ func AvailabilityZone(customObject v1alpha1.AWSConfig) string {
...
 }
 
-func AvailabilityZones(customObject v1alpha1.AWSConfig) int {
-	return customObject.Spec.AWS.AvailabilityZones
+func AvailabilityZones(customObject v1alpha1.AWSConfig) []v1alpha1.AWSConfigStatusAWSAvailabilityZone {
+	zones := customObject.Status.AWS.AvailabilityZones
+	sort.Slice(zones, func(i, j int) bool {
+		return zones[i].Name < zones[j].Name
+	})
+	return zones
```
Comment (xh3b4sd) [inline service/controller/v19/key/key.go]: When this is a thing the sorting should rather be done with the randomization directly in the ipam resource. When this is not a general thing we want it should may be only be done where we want exactly that behaviour.

File: service/controller/v19/key/key.go [hunk 5]
```diff
@@ -404,6 +409,10 @@ func RegionARN(customObject v1alpha1.AWSConfig) string {
 	return regionARN
 }
 
+func RequestedAvailabilityZones(customObject v1alpha1.AWSConfig) int {
```
Comment (xh3b4sd) [inline service/controller/v19/key/key.go]: Can we try to establish some scheme here? Why not put spec and status in the function names?

Comment (tuommaki) [reply service/controller/v19/key/key.go]: Renamed them to `{Spec,Status}AvailabilityZones()`

File: service/controller/v19/key/key.go (hunk 4 above)
Comment (tuommaki) [reply service/controller/v19/key/key.go]: I moved sorting to site-of-use.

**Pull request merged**